Rubberduck.UnitTesting/UnitTesting/VBEInteraction.cs

```diff
@@ -1,23 +1,68 @@
-﻿using System;
-using System.Runtime.InteropServices;
-using System.Reflection;
+﻿
 using Microsoft.Win32;
+using NLog;
+using Rubberduck.Parsing.Symbols;
 using Rubberduck.Resources.Registration;
+using Rubberduck.VBEditor.ComManagement.TypeLibs;
+using Rubberduck.VBEditor.ComManagement.TypeLibsAPI;
 using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
 
 namespace Rubberduck.UnitTesting
-{
-    [ComVisible(false)]
-    public static class ProjectTestExtensions
+{ 
+    // FIXME litter some logging around here
+    public class VBEInteraction
     {
-        public static void EnsureReferenceToAddInLibrary(this IVBProject project)
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+        internal static void RunDeclarations(IVBETypeLibsAPI typeLibApi, ITypeLibWrapper typeLib, IEnumerable<Declaration> declarations)
         {
+            foreach (var declaration in declarations)
+            {
+                typeLibApi.ExecuteCode(typeLib, declaration.QualifiedModuleName.ComponentName,
+                    declaration.QualifiedName.MemberName);
+            }
+        }
+
+        internal static void RunTestMethod(IVBETypeLibsAPI tlApi, ITypeLibWrapper typeLib, TestMethod test, EventHandler<AssertCompletedEventArgs> assertCompletionHandler, out long duration)
+        {
+            AssertHandler.OnAssertCompleted += assertCompletionHandler;
+            var stopwatch = new Stopwatch();
+            try
+            {
+                var testDeclaration = test.Declaration;
+
+                stopwatch.Start();
+                tlApi.ExecuteCode(typeLib, testDeclaration.ComponentName, testDeclaration.QualifiedName.MemberName);
+                stopwatch.Stop();
+
+                duration = stopwatch.ElapsedMilliseconds;
+            }
+            catch (Exception)
+            {
+                stopwatch.Stop();
+                duration = stopwatch.ElapsedMilliseconds;
+                throw;
+            }
+            finally
+            {
+                AssertHandler.OnAssertCompleted -= assertCompletionHandler;
+            }
+        }
+            
+
+        public static void EnsureProjectReferencesUnitTesting(IVBProject project)
+        {
+            if (project == null || project.IsWrappingNullReference) { return; }
             var libFolder = IntPtr.Size == 8 ? "win64" : "win32";
-            // TODO: This assumes the current assembly is same major/minor as the TLB!!!
+            // FIXME: This assumes the current assembly is same major/minor as the TLB!!!
```

Comment:
So, needs to be `RubberduckGuid.RubberduckUnitTestingTypeLibGuid`?

Reply:
Only if it should contain any COM-visible types. I don't think it should, though?

(Memo to myself - need to ensure there's an analyzer rule for checking the assembly's guid when it contains a COM-visible type)

Reply:
No. That's not where the COM types for unit testing reside. The COM types for unit-testing are in Rubberduck.Main (aka. Rubberduck). As such the Guid is correct. What's not correct is the way the Version is determined.

To cut a long story short: We basically need a way to inject a `Version` somewhere that's controlled by CW to get the Version reported by the Rubberduck.Main assembly.

This can be achieved by turning the VBEInteraction into a component that is controlled by CW. I do not much like that, though, because that automatically implies hiding that behind an interface and mocking that for testing. It also expands the number of dependencies that are put into the TestEngine.

A possible solution to alleviate this would be trying to push more of the IVBETypeLib interactions into the VBEInteraction class, pushing some dependencies with it. I'm not sure how feasible that is, though.

It works for the IVBETypeLibApi in and of itself, so that's one step taken. The Vbe design mode check could also be put in there, though at this point that's only because the names are related. I'll need to muck around a bit to properly resolve this...